src/services/admin/AuditLogsService.ts

```diff
@@ -0,0 +1,500 @@
+                        )
+                        return new DatabaseError(
+                          "Error occurred while retrieving review request data from the database"
+                        )
+                      }
+                    ).map((reviewRequest) => ({
+                      timestamp: new Date(pull.merged_at || ""),
+                      activity: AuditableActivity.PublishedChanges,
+                      actor:
+                        "requestor" in reviewRequest &&
+                        reviewRequest.requestor.email
+                          ? reviewRequest.requestor.email
+                          : pull.user?.login || "Unknown user",
+                      page: "",
+                      remarks: `GitHub Pull Request ID #${pull.number}`,
+                    }))
+                  )
+                )
+              )
+          )
+          .andThen((publishedChanges) =>
+            okAsync(
+              publishedChanges
+                .concat(auditLogs)
+                .sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime())
+            )
+          )
+      )
+  }
+
+  getAuditLogsViaFormsg(
+    email: string,
+    repoNames: string[],
+    sinceDate: string,
+    untilDate: string,
+    formSubmissionId: string
+  ): ResultAsync<void, AuditLogsError | ForbiddenError> {
+    const emailSubject = `[Isomer] Site Audit Logs (submission ID: ${formSubmissionId})`
+    logger.info(
+      `Received request to get audit logs, submission ID: ${formSubmissionId}`
+    )
+
+    // Step 1: Check if the user exists
+    return ResultAsync.fromPromise(this.usersService.findByEmail(email), () => {
+      logger.error(
+        `Error occurred while retrieving user data from the database for email ${email}`
+      )
+
+      return new DatabaseError(
+        "Error occurred while retrieving user data from the database"
+      )
+    })
+      .andThen((user) => {
+        if (!user) {
+          logger.warn(`Email address ${email} is not registered on Isomer CMS`)
+
+          return errAsync(
+            new ForbiddenError("Email address is not registered on Isomer CMS")
+          )
+        }
+
+        return okAsync(user)
+      })
+      .andThen((user) =>
+        // Step 2: Check if the user is a collaborator of ALL given repos
+        ResultAsync.combine([
+          okAsync(user),
+          ...repoNames.map((repoName) =>
+            ResultAsync.fromPromise(
+              this.collaboratorsService.getRole(repoName, user.id.toString()),
+              () =>
+                new DatabaseError(
+                  `Error retrieving user's role for the repo ${repoName}`
+                )
+            )
+          ),
+        ])
+      )
+      .andThen(([user, ...roles]) => {
+        const isUserNotAdminCollaborator = roles.some(
+          (role) =>
+            !role ||
+            (role !== CollaboratorRoles.Admin &&
+              role !== CollaboratorRoles.IsomerAdmin)
```

Comment:
hmm i thought we are gating this to devs only? shouldnt this just be isomerAdmin?

Reply:
Oops I meant this form is going to be used by agencies themselves as it is supposed to be self-service. For a start, we will be the ones using it (since they don't know the existence of the form) but the main intent is for them to use.